algorithms/src/Kokkos_Random.hpp

```diff
@@ -1800,30 +1569,62 @@ struct fill_random_functor_begin_end<ViewType, RandomPool, loops, 8,
   }
 };
 
-}  // namespace Impl
-
 template <class ViewType, class RandomPool, class IndexType = int64_t>
 void fill_random(ViewType a, RandomPool g,
-                 typename ViewType::const_value_type range) {
+                 typename ViewType::const_value_type begin,
+                 typename ViewType::const_value_type end) {
   int64_t LDA = a.extent(0);
   if (LDA > 0)
     parallel_for("Kokkos::fill_random", (LDA + 127) / 128,
-                 Impl::fill_random_functor_range<ViewType, RandomPool, 128,
-                                                 ViewType::Rank, IndexType>(
-                     a, g, range));
+                 Impl::fill_random_functor_begin_end<ViewType, RandomPool, 128,
+                                                     ViewType::Rank, IndexType>(
+                     a, g, begin, end));
 }
 
+}  // namespace Impl
+
 template <class ViewType, class RandomPool, class IndexType = int64_t>
 void fill_random(ViewType a, RandomPool g,
                  typename ViewType::const_value_type begin,
                  typename ViewType::const_value_type end) {
-  int64_t LDA = a.extent(0);
-  if (LDA > 0)
-    parallel_for("Kokkos::fill_random", (LDA + 127) / 128,
-                 Impl::fill_random_functor_begin_end<ViewType, RandomPool, 128,
-                                                     ViewType::Rank, IndexType>(
-                     a, g, begin, end));
+  switch (rank(a)) {
+    case 0:
+      Impl::fill_random(Impl::as_view_of_rank_n<0>(a), g, begin, end);
+      break;
+    case 1:
+      Impl::fill_random(Impl::as_view_of_rank_n<1>(a), g, begin, end);
+      break;
+    case 2:
+      Impl::fill_random(Impl::as_view_of_rank_n<2>(a), g, begin, end);
+      break;
+    case 3:
+      Impl::fill_random(Impl::as_view_of_rank_n<3>(a), g, begin, end);
+      break;
+    case 4:
+      Impl::fill_random(Impl::as_view_of_rank_n<4>(a), g, begin, end);
+      break;
+    case 5:
+      Impl::fill_random(Impl::as_view_of_rank_n<5>(a), g, begin, end);
+      break;
+    case 6:
+      Impl::fill_random(Impl::as_view_of_rank_n<6>(a), g, begin, end);
+      break;
+    case 7:
+      Impl::fill_random(Impl::as_view_of_rank_n<7>(a), g, begin, end);
+      break;
+    default:
+      Kokkos::Impl::throw_runtime_exception(
+          "Calling fill_random with a view of unexpected rank " +
+          std::to_string(rank(a)));
```

Comment:
Seeing this `switch(rank(a))`  form repeated in a few places now (both in Kokkos and downstream EMPIRE code) makes me think we should add something like

```c++
template <typename ViewType, typename Function>
void apply_to_view_of_static_rank(ViewType a, Function f)
{
  switch(rank(a)) {
  case 0: f(Impl::as_view_of_rank<0>(a); break;
  case 1: f(Impl::as_view_of_rank<1>(a); break;
  case 2: f(Impl::as_view_of_rank<2>(a); break;
  // etc
  }
}
```

Which would then make the call

```c++
apply_to_view_of_static_rank(a, [&] (auto a) { Impl::fill_random(a, g, begin, end); });
```

If I understand correctly, that generic lambda generates a functor with a templated `operator()`, so the compiler should do the right thing with nothing more spelled out.

I'd need to check the standard wording whether it should be called `apply_to_...` or `invoke_on_...`

Reply:
Would it make sense to implement and use that function in this PR?

Reply:
https://github.com/masterleinad/kokkos/pull/5 There's a PR to do just that.